---
title: "INFIMA: Integrative Fine-Mapping with Model Organism Multi-Omics Data"
author:
- name: Chenyang Dong
  affiliation: 
  - Department of Statistics, University of Wisconsin - Madison
- name: Sunduz Keles
  affiliation: 
  - Department of Biostatistics and Medical Informatics, University of Wisconsin - Madison
  - Department of Statistics, University of Wisconsin - Madison
date: "`r Sys.Date()`"
output: BiocStyle::html_document
vignette: >
  %\VignetteIndexEntry{INFIMA}
  %\VignetteEncoding{UTF-8}
  %\VignetteEngine{knitr::rmarkdown}
editor_options: 
  chunk_output_type: console
---

```{r eval=TRUE, echo=FALSE}
suppressWarnings(library(knitr))
```


# Introduction

This document provides an introduction to the *In*tegrative *Fi*ne-*Ma*pping with Model Organism Multi-Omics Data using the *INFIMA* package. *INFIMA* utilizes the diversity outbred (DO) mice population as the model organsim. The major usage of the *INFIMA* package is to fine-map the eQTL markers in DO studies (DO-eQTL). *INFIMA* implements an empirical Bayes model which quantifies how well each non-coding SNP explains the observed DO allelic patern through consistency of founder mice RNA-seq data, founder mice ATAC-seq data (including the existence and consistency of a footprint) with the observed allelic pattern. *INFIMA* assumes 0 or 1 causal SNPs for a DO-eQTL gene. Given a library of mouse SNPs, the corresponding local ATAC-seq signals, DO-eQTL data as well as the gene expression in founder mice, *INFIMA* provides the following functionalities:

(a) Fine-mapping DO-eQTLs and estimate SNP-level posterior probabilities.
(b) Linking SNPs or the local ATAC-seq peaks to effector genes.

*INFIMA* results from DO studies can be mapped to human, which provides putative effector genes of human GWAS SNPs.

# Installation

INFIMA will be submitted to Bioconductor. Currently, INFIMA can be downloaded and installed in R by: 

```{r eval=FALSE}
devtools::install_github("ThomasDCY/INFIMA")
```


*INFIMA* depends on the following \R{} packages:

(a) `r CRANpkg("data.table")` is used for formatting results that are easy for users to query.
(b) `r Biocpkg("GenomicRanges")` is used for operating genomic intervals.
(c) `r CRANpkg("doParallel")` is used for parallel computing.

# Example

## Input data

In this section, we will illustrate how to use *INFIMA* with a simple example containing 10 DO genes in chr1.

```{r eval=TRUE}
library(INFIMA)
data('example-10-genes')
```

We have three pieces of data in total.

### DO-eQTL data

The DO-eQTL data frame should contain the following columns:

name          | explanation
------------- | ----------------------------------------
ensembl       | The ensembl ID of DO-eQTL gene.
symbol        | The symbol of DO-eQTL gene.
chr           | The chromosome of DO-eQTL gene.
start         | The start position of DO-eQTL gene.
end           | The end position of DO-eQTL gene.
strand        | The strand of DO-eQTL gene.
qtl_marker    | The ID of DO-eQTL marker.
qtl_chr       | The chromosome of DO-eQTL marker.
qtl_pos       | The position of DO-eQTL marker.
lod           | The LOD score of DO-eQTL marker.

Plus 8 columns named 129, AJ, B6, CAST, NOD, NZO, PWK, WSB representing the DO-eQTL allelic effect sizes for the 8 DO founder strains.
```{r}
dt1[1,] # DO-eQTL data
```


### Founder RNA-seq data

The founder RNA-seq data is a data frame object which is essentially a DO gene (row) by sample (column) matrix with normalized gene expressions.
The column names must be strain names (129, AJ, B6, CAST, NOD, NZO, PWK, WSB)
and there must be at least one sample for each strain.
One caveat is that the rows of founder RNA-seq data should
correspond to the rows of DO-eQTL data (the same DO gene), i.e. the first two input files
have the same number of rows.
```{r}
nrow(dt1) == nrow(dt2) # the genes should be the same
```


### SNP and local ATAC-seq signal data

The SNP and local ATAC-seq signal data is a data frame containing the following columns
with exactly the same ordering:

name          | explanation
------------- | ----------------------------------------
snp_id        |  The SNP ID. 
chr           |  The chromosome of the SNP. 
snp_pos       |  The position of the SNP. 
ref           |  The reference nucleotide. 
alt           |  The alternative nucleotide. 
footprint     |  The user-defined footprint annotation.

The recommended range for the footprint column is between 0 and 1. The larger the value,
the more confidence the SNP is affecting a footprint. The maximum value can exceed 1 if the user wants to emphasize the importance of footprint analysis. Put 0 if the footprint analysis is not applicable. In our application, we tagged the local-ATAC-QTLs (SNPs best correlate with local ATAC-seq signals) affecting footprints as 1 and 0 otherwise.

Plus

(a) 8 columns named as strain-genotype where strain
is 129, AJ, B6, CAST, NOD, NZO, PWK, WSB in order. 0 means reference
nucleotide while 2 means alternative nucleotide.

(b) 8 columns named as strain-ATAC-seq where strain
is 129, AJ, B6, CAST, NOD, NZO, PWK, WSB in order. The normalized
local ATAC-seq signal of the SNP.

```{r}
dt3[1,]
```


## Process the input data

```{r}
raw_data <- raw_input_data(dt1, dt2, dt3)
model_data <- model_input_data(raw_data)
```

## Compute the Dirichlet prior

```{r}
prior <- compute_prior(raw_data, model_data)
```

## INFIMA model fitting

```{r}
infima <- model_fitting(model_data, prior)
```

## Analysis of the results

```{r}
infima_results <- snp_link_gene(infima, raw_data, model_data, prior, fdr = 0.05, cum.pprob = 0.8, cred.set = 0.5)
summary(infima_results)
```



# Session Information

```{r eval=TRUE,echo=FALSE,results="markup",cache=FALSE}
print(sessionInfo())
```